ArmPkg/Library/TpmServiceLib/TpmServiceLib.c

```diff
@@ -0,0 +1,333 @@
+TpmStatus
+GetInterfaceVersionHandler (
+  DIRECT_MSG_ARGS_EX *Request, 
+  DIRECT_MSG_ARGS_EX *Response
+  )
+{
+  (void)Request;
+  TpmStatus ReturnVal = TPM_STATUS_OK;
+  Response->Arg0 = TPM_STATUS_OK_RESULTS;
+  Response->Arg1 = (TPM_MAJOR_VER << 16) | TPM_MINOR_VER;
+  return ReturnVal;
+}
+
+STATIC
+TpmStatus
+GetFeatureInfoHandler (
+  DIRECT_MSG_ARGS_EX *Request, 
+  DIRECT_MSG_ARGS_EX *Response
+  )
+{
+  (void)Request;
+  TpmStatus ReturnVal = TPM_STATUS_OK;
+  Response->Arg0 = TPM_STATUS_NOT_SUP;
+  DEBUG ((DEBUG_ERROR, "Unsupported Function\n"));
+  return ReturnVal;
+}
+
+STATIC
+TpmStatus
+StartHandler (
+  DIRECT_MSG_ARGS_EX *Request, 
+  DIRECT_MSG_ARGS_EX *Response
+  )
+{
+  TpmStatus ReturnVal = TPM_STATUS_OK;
+  UINT8 Function = Request->Arg1;
+  UINT8 Locality = Request->Arg2;
+
+  /* Based on the function we will need to read from the corresponding address
+    * register in the tpm_crb to know where to pull the data from: 
+    * NOTE: function = 0, command is ready to be processed
+    *       function = 1, locality request is ready to be processed
+    *       locality = 0...4, the locality where the command or request is located */
+  /* TODO: Currently only Locality0 is available, if this needs to change, we should
+   *       update the PCD base address with the appropriate locality offset. */
+  if (Locality != PTP_CRB_LOCALITY_STATE_ACTIVE_LOCALITY_0) {
+    Response->Arg0 = TPM_STATUS_INVARG;
+    DEBUG ((DEBUG_ERROR, "Invalid Locality\n"));
+    return TPM_STATUS_INVARG;
+  }
+  
+  if (Function == TPM_START_PROCESS_CMD) {
+      ReturnVal = HandleCommand();
+  } else if (Function == TPM_START_PROCESS_LOC_REQ) {
+      ReturnVal = HandleLocalityRequest();
+  } else {
+      ReturnVal = TPM_STATUS_INVARG;
+      DEBUG ((DEBUG_ERROR, "Invalid Start Function\n"));
+  }
+
+  /* Clean up the internal CRB at the given locality */
+  CleanInternalCrb();
+
+  Response->Arg0 = ReturnVal;
+  return ReturnVal;
+}
+
+STATIC
+TpmStatus
+RegisterHandler (
+  DIRECT_MSG_ARGS_EX *Request, 
+  DIRECT_MSG_ARGS_EX *Response
+  )
+{
+  (void)Request;
+  TpmStatus ReturnVal = TPM_STATUS_OK;
+  Response->Arg0 = TPM_STATUS_NOT_SUP;
+  DEBUG ((DEBUG_ERROR, "Unsupported Function\n"));
+  return ReturnVal;
+}
+
+STATIC
+TpmStatus
+UnregisterHandler (
+  DIRECT_MSG_ARGS_EX *Request, 
+  DIRECT_MSG_ARGS_EX *Response
+  )
+{
+  (void)Request;
+  TpmStatus ReturnVal = TPM_STATUS_OK;
+  Response->Arg0 = TPM_STATUS_NOT_SUP;
+  DEBUG ((DEBUG_ERROR, "Unsupported Function\n"));
+  return ReturnVal;
+}
+
+STATIC
+TpmStatus
+FinishHandler (
+  DIRECT_MSG_ARGS_EX *Request, 
+  DIRECT_MSG_ARGS_EX *Response
+  )
+{
+  (void)Request;
+  TpmStatus ReturnVal = TPM_STATUS_OK;
+  Response->Arg0 = TPM_STATUS_NOT_SUP;
+  DEBUG ((DEBUG_ERROR, "Unsupported Function\n"));
+  return ReturnVal;
+}
+
+VOID
+TpmServiceInit (
+  VOID
+  )
+{
+  InitInternalCrb();
+}
+
+VOID
+TpmServiceDeInit (
+  VOID
+  )
+{
+  /* Nothing to DeInit */
+}
+
+VOID
+TpmServiceHandle (

```

Comment:
I think these libraries are well architected to be in separate commits.

Let's break this PR into 3 commits (maybe 4 if we stick to the timer lib you have here), one for each service.